src/alpaka/AlpakaCore/CachingAllocator.h

```diff
@@ -311,11 +316,24 @@ namespace cms::alpakatools {
       return false;
     }
 
+    Buffer allocateBuffer(size_t bytes, Queue const& queue) {
+      if constexpr (std::is_same_v<Device, alpaka::Dev<Queue>>) {
+        // allocate device memory
+        return alpaka::allocBuf<std::byte, size_t>(device_, bytes);
+      } else if constexpr (std::is_same_v<Device, alpaka::DevCpu>) {
+        // allocate pinned host memory
+        return alpaka::allocMappedBuf<std::byte, size_t>(device_, alpaka::getDev(queue), bytes);
+      } else {
+        // unsupported combination
+        static_assert(std::is_same_v<Device, alpaka::Dev<Queue>> or std::is_same_v<Device, alpaka::DevCpu>,
```

Comment:
IIUC, this `static_assert` does not add anything to the one on line 93, so in principle the `else` branch is redundant.

Reply:
That's correct.

However, I didn't feel comfortable with
```c++
      if constexpr (/* device allocator check */) { 
        // allocate device memory
        return ...
      } else if constexpr (/* host allocator check */) { 
        // allocate pinned host memory
        return ...
      } 
```
without any `else` branch, even if the `static_assert` on line 93 prevents the other cases.

And I liked even less making the second condition implicit, like in
```c++
      if constexpr (/* device allocator check */) { 
        // allocate device memory
        return ...
      } else  { 
        // allocate pinned host memory
        return ...
      } 
```
because a change to the `static_assert` above would change the behaviour of the `else` branch, which may not be a clear side effect.

In the end repeating the `static_assert` seemed the least bad solution 🤷

Reply:
I agree it makes the intention more clear to follow.(and any runtime error would feel more silly), so I'm fine to keep it.